daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala

```diff
@@ -254,71 +244,69 @@ final class ImmArray[+A] private (
     * return the `array` as is and people would be able to break the original
     * `ImmArray`.
     */
-  def toSeq: ImmArray.ImmArraySeq[A] = new ImmArray.ImmArraySeq[A](this)
+  def toSeq: ImmArray.ImmArraySeq[A] = toIndexedSeq
 
   /** O(n) */
-  def collect[B](f: PartialFunction[A, B]): ImmArray[B] = {
-    val builder = ImmArray.newBuilder[B]
-    var i = 0
-    while (i < length) {
-      val a = uncheckedGet(i)
-      if (f.isDefinedAt(a)) builder += f(a)
-      i += 1
-    }
-    builder.result()
-  }
+  def collect[B](f: PartialFunction[A, B]): ImmArray[B] =
+    ImmArray.fromIterator(iterator.collect(f))
 
   /** O(n) */
   def foreach(f: A => Unit): Unit = {
-    @tailrec
-    def go(cursor: Int): Unit =
-      if (cursor < length) {
-        f(uncheckedGet(cursor))
-        go(cursor + 1)
-      }
-    go(0)
+    val end = start + length
+    var i = start
+    while (i < end) {
+      f(array(i).asInstanceOf[A])
+      i += 1
+    }
   }
 
   /** O(n) */
   def foldLeft[B](z: B)(f: (B, A) => B): B = {
-    @tailrec
-    def go(cursor: Int, acc: B): B = {
-      if (cursor < length)
-        go(cursor + 1, f(acc, uncheckedGet(cursor)))
-      else
-        acc
+    val end = start + length
+    var acc = z
+    var i = start
+    while (i < end) {
+      acc = f(acc, array(i).asInstanceOf[A])
+      i += 1
     }
-    go(0, z)
+    acc
   }
 
   /** O(n) */
   def foldRight[B](z: B)(f: (A, B) => B): B = {
-    @tailrec
-    def go(cursor: Int, acc: B): B =
-      if (cursor >= 0)
-        go(cursor - 1, f(uncheckedGet(cursor), acc))
-      else
-        acc
-    go(length - 1, z)
+    var i = length - 1
+    var acc = z
+    while (i >= start) {
+      acc = f(array(i).asInstanceOf[A], acc)
+      i -= 1
+    }
```

Comment:
I know there are no unsigned integers in Scala, but still: 😃
```suggestion
    var i = length
    var acc = z
    while (i > start) {
      i -= 1
      acc = f(array(i).asInstanceOf[A], acc)
    }
```